src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -1220,102 +1222,876 @@ private static float MinMaxCore<TMinMaxOperator>(ReadOnlySpan<float> x)
+
+                    Vectorized512Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+#endif
+
+            if (Vector256.IsHardwareAccelerated)
+            {
+                if (remainder >= (uint)(Vector256<float>.Count))
+                {
+                    Vectorized256(ref xRef, ref dRef, remainder);
+                }
+                else
+                {
+                    // We have less than a vector and so we can only handle this as scalar. To do this
+                    // efficiently, we simply have a small jump table and fallthrough. So we get a simple
+                    // length check, single jump, and then linear execution.
+
+                    Vectorized256Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+
+            if (Vector128.IsHardwareAccelerated)
+            {
+                if (remainder >= (uint)(Vector128<float>.Count))
+                {
+                    Vectorized128(ref xRef, ref dRef, remainder);
+                }
+                else
+                {
+                    // We have less than a vector and so we can only handle this as scalar. To do this
+                    // efficiently, we simply have a small jump table and fallthrough. So we get a simple
+                    // length check, single jump, and then linear execution.
+
+                    Vectorized128Small(ref xRef, ref dRef, remainder);
+                }
+
+                return;
+            }
+
+            // This is the software fallback when no acceleration is available
+            // It requires no branches to hit
+
+            SoftwareFallback(ref xRef, ref dRef, remainder);
+
+            static void SoftwareFallback(ref float xRef, ref float dRef, nuint length)
+            {
+                for (nuint i = 0; i < length; i++)
+                {
+                    Unsafe.Add(ref dRef, i) = TUnaryOperator.Invoke(Unsafe.Add(ref xRef, i));
+                }
+            }
+
+            static void Vectorized128(ref float xRef, ref float dRef, nuint remainder)
+            {
+                ref float dRefBeg = ref dRef;
+
+                // Preload the beginning and end so that overlapping accesses don't negatively impact the data
+
+                Vector128<float> beg = TUnaryOperator.Invoke(Vector128.LoadUnsafe(ref xRef));
+                Vector128<float> end = TUnaryOperator.Invoke(Vector128.LoadUnsafe(ref xRef, remainder - (uint)(Vector128<float>.Count)));
+
+                if (remainder > (uint)(Vector128<float>.Count * 8))
+                {
+                    // Pinning is cheap and will be short lived for small inputs and unlikely to be impactful
+                    // for large inputs (> 85KB) which are on the LOH and unlikely to be compacted.
+
+                    fixed (float* px = &xRef)
+                    fixed (float* pd = &dRef)
```

Comment:
What's the thinking for using fixed and pointers instead of refs? The same situation that results in alignment changing is also the one that results in pinning being expensive.

Reply:
Without pinning you then can not use APIs like `StoreAlignedNonTemporal`

Reply:
Using fixed is required to support non-temporal stores, which significantly improves perf for large inputs.

Pinning itself is cheap, it is just spilling an address to the stack. The potential impact of pinning is that it can negatively impact the GC's ability to perform compaction of memory.

For "small" data, we avoid pinning entirely. Where small is currently `>= 128` bytes for V128, `>= 256` bytes for V256, `>= 512` bytes for V512.

For medium sized data (`<= 85KB`, so on the regular GC heap), the pin is going to be short and the biggest risk is if we repeatedly pin the data across many calls.

For large sized data (`> 85KB`, on the LOH), allocations are already unlikely to be compacted and so the pinning shouldn't be an issue.

Reply:
Ah, ok. As part of making this non-draft, it'd be nice if the comment explained that.

Reply:
> The potential impact of pinning is that it can negatively impact the GC's ability to perform compaction of memory.

Right, that's why I was saying the situation that would cause the negative impact with pinning is the same situation that would cause the negative impact with assuming a particular alignment with refs (since you could all of a sudden become unaligned).

Egor's explanation about StoreAlignedNonTemporal explains the rationale.

Reply:
> Where small is currently `>= 128` bytes for V128, `>= 256` bytes for V256, `>= 512` bytes for V512.

@tannergooding you have typos in here, should be `<=`.